backupccl: add test with randomized completed spans to TestRestoreEntryCover

rhu713 opened the pull request:
Add some testing with randomized completed spans to TestRestoreEntryCover. This testing should demonstrate the correctness of generateAndSendImportSpans in the presence of completed spans.

Informs: #98779

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/99304)
<!-- Reviewable:end -->

msbutler commented on pkg/ccl/backupccl/restore_span_covering_test.go:
```diff
@@ -270,7 +271,7 @@ func makeImportSpans(
+		highWaterMark,
 		introducedSpanFrontier,
 		targetSize,
 		true)
```
the highWaterMark will not get exercised unless that last arg that is currently set to `true` is changed. We should set it to false if the HWM is not empty.

msbutler commented on pkg/ccl/backupccl/restore_span_covering_test.go:
```diff
@@ -942,13 +971,65 @@ func TestRestoreEntryCover(t *testing.T) {
+											completedSpans = append(completedSpans, roachpb.Span{
+												Key:    cover[0].Span.Key,
+												EndKey: cover[idx].Span.EndKey,
+											})
```
ack! This made me realize there's a bad bug if `bulkio.restore.use_simple_import_spans` is true on 23.1, where we currently only use the hwm to checkpoint things. But the hwm version of checkpointing is _version_ gated not _feature_ gated. This implies that if use SimpleImportSpans, we don't get any checkpointing goodness.

I'll send out a patch.

msbutler replied:
We should merge [this](https://github.com/cockroachdb/cockroach/pull/99373) before this pr.

Commit:
backupccl: add test with randomized completed spans to TestRestoreEntryCover

Add some testing with randomized completed spans to
TestRestoreEntryCover. This testing should demonstrate the correctness
of generateAndSendImportSpans in the presence of completed spans.

Informs: #98779

Release note: None

msbutler commented on pkg/ccl/backupccl/restore_span_covering_test.go:
```diff
@@ -1009,13 +1042,81 @@ func TestRestoreEntryCover(t *testing.T) {
+											}
+										} else {
+											idx := r.Intn(len(cover))
+											completedSpans = append(completedSpans, roachpb.Span{
```
nit: we shouldn't need to create a list of completed spans for this to work. If restore is still using a hwm, the complete spans list will always be empty.

msbutler reviewed (approved):
left a few nits

rhu713 commented:
bors r+
